crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -1368,6 +1390,50 @@ pub async fn make_warrant_op_inner(
     Ok(op)
 }
 
+pub async fn make_fork_warrant_op_inner(
+    keystore: &MetaLairClient,
+    warrant_author: AgentPubKey,
+    chain_author: AgentPubKey,
+    action_pair: ((ActionHash, Signature), (ActionHash, Signature)),
+) -> WorkflowResult<DhtOpHashed> {
+    tracing::warn!("Authoring warrant for chain fork by {chain_author}");
+
+    let warrant = Warrant::ChainIntegrity(ChainIntegrityWarrant::ChainFork {
+        chain_author,
+        action_pair,
+    });
+    let warrant_op = WarrantOp::author(keystore, warrant_author, warrant)
+        .await
+        .map_err(|e| super::WorkflowError::Other(e.into()))?;
+    let op: DhtOp = warrant_op.into();
+    let op = op.into_hashed();
+    Ok(op)
+}
+
+pub fn detect_fork(
+    txn: &mut Transaction<'_>,
+    action: &Action,
+) -> StateQueryResult<Option<(ActionHash, Signature)>> {
+    let mut statement = txn.prepare(ACTION_HASH_BY_PREV)?;
+    statement
+        .query_row(
+            named_params! {
+                ":prev_hash": action.prev_action(),
+                ":author": action.author()
```

Comment:
Will this have to be updated with key updates, so that I cannot rotate my key and then create a fork?

Reply:
This was phrased wrongly, I mean can I rotate my key and then create a fork and it will not be noticed because the query looks for the new agent and the other actions were written with the previous agent?

Reply:
Egad, you're right! The query has to have a conditional, where if it's an Update action, it checks the updated key rather than the author, because the Update will be authored by the old author.

Reply:
@jost-s could you review my handling of this? https://github.com/holochain/holochain/pull/3996/commits/98585c8e30e7de401647f80d73a1cea24d328432